proposals/nnnn-inferring-senable-for-methods.md

```diff
@@ -0,0 +1,221 @@
+
+Next is:
+
+3. The inference of `@Sendable`  when referencing non-local functions.
+
+Unlike closures, which retain the captured value, global functions can't capture any variables - because global variables are just referenced by the function without any ownership. With this in mind there is no reason not to make these `Sendable` by default. This change will also include static global functions.
+
+```
+func doWork() -> Int {
+`  Int.random(in: 1..<42)`
+}
+
+Task<Int, Never>.detached(priority: **nil**, operation: doWork) // Converting non-sendable function value to '@Sendable () async -> Void' may introduce data races
+```
+
+Currently, trying to start a `Task` with the global function `doWork` will cause an error complaining that the function is not `Sendable`. This should compile with no issue.  
+
+4. Prohibition of marking methods `@Sendable` when the type they belong to is not `@Sendable`.
+```
+    class C {
+        var random: Int = 0 // random is mutable so `C` can't be checked sendable
+        
+        @Sendable func generateN() async -> Int { //error: adding @Sendable to function of non-Senable type prohibited
+             random = Int.random(in: 1..<100)
+             return random
+        }
+    }
+
+    func test(c: C) { c.generateN() }
+
+    let num = C()
+    Task.detached {
+       test(num)
+    }
+    test(num) // data-race
+```
+
+If we move the previous work we wanted to do into a class that stores the random number we generate as a mutable value, we could be introducing a data race by marking the function responsible for this work `@Sendable` . Doing this should be prohibited by the compiler.
+
+Since `@Sendable` attribute will be automatically determined with this proposal, you will no longer have to explicitly write it on function and method declarations.
+
+## Source compatibility
+
+No impact.
+
+## Effect on ABI stability
+
+When you remove an explicit `@Sendable` from a method, the mangling of that method will change. Since `@Sendable` will now be inferred, if you choose to remove the explicit annotation to "adopt" the inference, you may need to consider the mangling change.
+
+## Effect on API resilience
+
+No effect on ABI stability. 
+
+## Future Directions 
+
+Accessors are not currently allowed to participate with the `@Sendable` system in this proposal. It would be straight-forward to allow getters to do so in a future proposal if there was demand for this.
+
+## Alternatives Considered 
+
+Swift could forbid explicitly marking function declarations with the` @Sendable` attribute, since under this proposal there’s no longer any reason to do this.
```

Comment:
Even in this alternative, there actually would still be a case where we'd allow `@Sendable` on function decls: local functions.
```suggestion
Swift could forbid explicitly marking non-local function declarations with the` @Sendable` attribute, since under this proposal there’s no longer any reason to do this.
```

Reply:
Can you explain a bit why we would we need to add @Sendable to local functions?

Reply:
chatted offline, @beccadax was referring to nested functions :) added the suggested change and updated proposal with an example